### PR TITLE
Preliminary version of AOD producer workflow

### PR DESCRIPTION
**Pull request by nburmaso**

### CODE REVIEW

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (sawenzel) [inline Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:53]: shouldn't this be defined as part of the data model (or put into if not present)?

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (sawenzel) [inline Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h]: Types start with a capital letter (even template ones).

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (sawenzel) [inline Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: The coding and naming conventions ask to put braces even around single line bodies. Can you please change this consistently throughout this PR? (There is an ongoing campaign to fix this in all of O2 but better to avoid these changes in the first place).

**Review by sawenzel — CHANGES_REQUESTED** [review]
Good starting point! Could you please address the cosmetic changes? 

Also, it would be important to have an AOD writer (writing the ROOT file) as a device. It does not seem to be part of this PR ... or is this provided automatically by the framework??

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (nburmaso) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:53]: As far as I understood the `AnalysisDataModel.h`, there is no `DECLARE_SOA_TABLE` for these columns. I tried joining `StoredTracks` and `StoredTracksCov` tables, but it did not work, since they contain expression and dynamic columns. Perhaps, I just did not understand how to do this right

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (nburmaso) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h]: Thank you, I will fix it

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx
Comment (nburmaso) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx]: Sure, I will add the brackets

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (sawenzel) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:53]: @jgrosseo @aalkin : Could you comment on this?

Comment (aalkin) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:53]: This is true, `StoredTracks` and `StoredTracksCov` cannot be joined 'as is'. This definition should be put in data model, as a workaround for now, then I'll make it properly depend on existing definitions.

Comment (jgrosseo) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:53]: We can use defines not to repeat the list of columns for each case, but it is not so nice.

**Comment by ktf** [comment]
@sawenzel yes, the framework will do the saving automatically. From its point of view this is no different than a derived table / skim.

File: Detectors/AOD/src/AODProducerWorkflow.cxx
Comment (ktf) [inline Detectors/AOD/src/AODProducerWorkflow.cxx]: why not:

```
framework::WorkflowSpec specs {
  o2::vertexing::getPrimaryVertexingSpec(useFT0, useMC),
  o2::globaltracking::getTrackTPCITSReaderSpec(useMC),
  o2::its::getITSTrackReaderSpec(useMC),
  o2::tpc::getTPCTrackReaderSpec(useMC),
  o2::ft0::getDigitReaderSpec(useMC),
  o2::ft0::getReconstructionSpec(useMC),
  o2::aodproducer::getAODProducerWorkflowSpec()
};
```

?

Comment (nburmaso) [reply Detectors/AOD/src/AODProducerWorkflow.cxx]: Yes, it would be better this way. I will change `specs` into this form.

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflow.h
Comment (ktf) [inline Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflow.h]: notice you can now (C++17) say:

```
namespace o2::aodproducer {
}
```

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (sawenzel) [inline Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h]: sorry to comment here again: Type notation starting with "_" is really not standard. Consider using `TracksT` or `TracksType` so that the type name really starts with a capital letter.

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflow.h
Comment (nburmaso) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflow.h]: Yes, this variant looks more neat.

File: Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h
Comment (nburmaso) [reply Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h]: Oh, I see. Sure, I will fix it.

**Review by sawenzel — DISMISSED** [review]
For me ok as a starting point. The Tracks table should still be moved into the data model definition at some point. @jgrosseo : I'll let you take care of the merge (preferably as one squashed commit).

**Comment by ktf** [comment]
Please make sure you fix the checkcode errors:

```
/mnt/mesos/sandbox/sandbox/o2-gpu/sw/SOURCES/O2/4637/0/Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:58:3: warning: use '= default' to define a trivial default constructor [modernize-use-equals-default]
  AODProducerWorkflowDPL() {}
  ^                        ~~
                           = default;
Suppressed 5124 warnings (5121 in non-user code, 3 NOLINT).
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
5416 warnings generated.
/mnt/mesos/sandbox/sandbox/o2-gpu/sw/SOURCES/O2/4637/0/Detectors/AOD/include/AODProducerWorkflow/AODProducerWorkflowSpec.h:58:3: warning: use '= default' to define a trivial default constructor [modernize-use-equals-default]
  AODProducerWorkflowDPL() {}
  ^                        ~~
                           = default;
```

**Comment by jgrosseo** [comment]
This is fine with me, except the checkcode errors
@sawenzel I will then take care of moving the table to the data model once this is merged

**Comment by ktf** [comment]
Given it's a separate workflow and now every check passes, I guess we can simply merge this and iterate.

**Pull request merged**
